versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/DatabaseAdapterExtension.java

```diff
@@ -0,0 +1,190 @@
+  private static final Namespace NAMESPACE = Namespace.create(DatabaseAdapterExtension.class);
+  private static final String KEY = "database-adapter";
+
+  /**
+   * Used to provide this extension an instance of {@link DatabaseAdapter}.
+   *
+   * @param context JUnit extension context, if needed
+   * @param testConfigurer {@link TestConfigurer}, must be passed though to the static {@code
+   *     createAdapter()} methods.
+   */
+  protected abstract DatabaseAdapter createAdapter(
+      ExtensionContext context, TestConfigurer testConfigurer);
+
+  @Target(ElementType.FIELD)
+  @Retention(RetentionPolicy.RUNTIME)
+  public @interface Adapter {
+    /**
+     * Allows per-test-class configuration for of database-adapter configurations using a {@link
+     * TestConfigurer}.
+     */
+    Class<? extends TestConfigurer> configurer() default NoopConfigurer.class;
+  }
+
+  /** Used to configure database-adapters per test-class. */
+  public interface TestConfigurer {
+    <C extends DatabaseAdapterConfig> DatabaseAdapterConfig configure(C config);
+  }
+
+  private static class NoopConfigurer implements TestConfigurer {
+    @Override
+    public <C extends DatabaseAdapterConfig> DatabaseAdapterConfig configure(C config) {
+      return config;
+    }
+  }
+
+  protected static <C extends DatabaseAdapterConfig> DatabaseAdapter createAdapter(
+      String adapterName,
+      TestConfigurer testConfigurer,
+      Function<C, DatabaseAdapterConfig> configurer) {
+    return createAdapter(
+        testConfigurer,
+        DatabaseAdapterFactory.<C>loadFactoryByName(adapterName).newBuilder(),
+        configurer);
+  }
+
+  protected static <C extends DatabaseAdapterConfig> DatabaseAdapter createAdapter(
+      TestConfigurer testConfigurer,
+      DatabaseAdapterFactory.Builder<C> adapterBuilder,
+      Function<C, DatabaseAdapterConfig> configurer) {
+    adapterBuilder =
+        adapterBuilder
+            .configure(SystemPropertiesConfigurer::configureFromSystemProperties)
+            // default to a quite small max-size for the CommitLogEntry.keyList + KeyListEntity
+            // This is necessary for AbstractManyKeys to work properly!!
+            .configure(c -> c.withMaxKeyListSize(2048))
+            .configure(configurer)
+            .configure(testConfigurer::configure);
+
+    return adapterBuilder.build();
+  }
+
+  static class AdapterResource implements CloseableResource {
+    private final DatabaseAdapter databaseAdapter;
+
+    AdapterResource(DatabaseAdapter databaseAdapter) {
+      this.databaseAdapter = databaseAdapter;
+    }
+
+    @Override
+    public void close() throws Throwable {
+      databaseAdapter.close();
+    }
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    Class<?> testClass = context.getRequiredTestClass();
+    findAnnotatedFields(testClass, Adapter.class, ReflectionUtils::isStatic)
```

Comment:
I'd do it differently. Since we assume the adapter is created in the `static` context we could declare the configurer in a class-level annotation.

This extension would them get the adapter name + configuration method from the annotation and always create the adapter through the factory. There's no need to subclass the extension itself. I think we can also get away without a `TestConfigurer`

For example:
```
@ExtendWith(DatabaseAdapterExtension.class)
@NessieAdapterConfig(createMethod="createAdapter")
public class TestMongoDatabaseAdapter extends AbstractTieredCommitsTest {
  static DatabaseAdapter createAdapter(...) {
    ...
  }
}
```


Then, we'd also implement `ParameterResolver` and inject the adapter into any/all lifecycle method parameters, be them static or per-instance.

`AbstractTieredCommitsTest` would have this method to receive the pre-configured adapter.
```
`@BeforeAll 
static void setup(DataAdapter adapter) {
  databaseAdapter = adapter;
}
```

WDYT?